Hash submitters in the deduplication key [DPP-347]

kamil-da opened the pull request:
## Motivation
Submitting a command with a large number of submitters results in:
```
org.postgresql.util.PSQLException: ERROR: index row size 6512 exceeds maximum 2712 for index "participant_command_submissions_pkey"
  Hint: Values larger than 1/3 of a buffer page cannot be indexed.
Consider a function index of an MD5 hash of the value, or use full text indexing.
```
due to how we construct deduplication keys. Concatenating a large number of submitters causes exceeding Postgres limits.

## Main changes
Made the deduplication key containing a hash of submitters instead of a concatenated string of submitters.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [x] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

rautenrieth-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala:
```diff
@@ -0,0 +1,56 @@
+        .update(
+          _.commands.deduplicationTime := deduplicationTime.asProtobuf
+        )
+      _ <- ledger.submit(request)
```
Shouldn't this test assert that the dummy contract was successfully created?

kamil-da replied:
Fixed 👍

rautenrieth-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala:
```diff
@@ -0,0 +1,56 @@
+
+final class ValueLimitsIT(timeoutScaleFactor: Double) extends LedgerTestSuite {
+
+  /** Postgres has a limit on the index row size of 2712.
+    * We test that a large submitters number doesn't cause deduplication failure because of that limit.
+    * THIS TEST CASE DOES NOT TEST ANY OTHER FAILURES THAT MAY BE CAUSED BY A LARGE SUBMITTERS NUMBER
+    */
```
This description is not accurate:
- The test uses the ledger API and not the internals of the indexer,  so you don't really know what kind of failures it tests for.
- The description mentions a limit of Postgres, but you don't know whether the test runs against a Postgres index database.
- You are not checking any property of command deduplication, you are just submitting a request with a large number of parties. I would name the test `"Submit commands with a large number of submitters"`.

kamil-da replied:
@rautenrieth-da Thank you for the review!
I've modified the test suite according to your suggestions. I've also added a change to a database index that was blocking submissions due to index row size limits.

rautenrieth-da reviewed (changes requested):
Any change in `com.daml.platform.store.dao` needs to be mirrored in `com.daml.platform.store.appendonlydao`, can you please edit `JdbcLedgerDao` there accordingly?

rautenrieth-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala:
```diff
@@ -0,0 +1,57 @@
+        .update(
+          _.commands.deduplicationTime := deduplicationTime.asProtobuf
+        )
+      _ <- ledger.submit(request)
+      _ <- Delayed.by(1.second)(())
```
```suggestion
      _ <- ledger.submitAndWait(request)
```

rautenrieth-da commented on ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala:
```diff
@@ -0,0 +1,57 @@
+          commands = DummyWithAnnotation(parties.head, "First submission").create.command,
+        )
+        .update(
+          _.commands.deduplicationTime := deduplicationTime.asProtobuf
```
Why do you need to set a custom deduplication time?

rautenrieth-da reviewed (changes requested):
Left some minor comments. You still need to do the change in `com.daml.platform.store.appendonlydao` though.

kamil-da commented:
@rautenrieth-da I've adapted the PR according to your review and the discussion during the meeting.
- implemented the feature for `appendonlydao` package
- removed changes to the database index - will come in a follow-up PR
- removed `ValueLimitsIT` suite - will come in a follow-up PR

For this PR I've replaced API-level testing with simple unit testing because it's currently impossible to make API-level tests pass without making changes to database indexes.

rautenrieth-da commented on ledger/participant-integration-api/src/test/suite/scala/platform/store/dao/DeduplicationKeyMakerSpec.scala:
```diff
@@ -0,0 +1,56 @@
+
+    "make a deduplication key with a limited length for a large number of submitters" in {
+      val submitters = (1 to 50).map(_ => aParty()).toList
+      val MaxKeyLength = 200
```
This constant seems arbitrary, but the value is fine. It's smaller than the maximum length for indexed string columns in both Postgres and Oracle (which was the reason for this change). Maybe add a comment?

kamil-da replied:
Good point - I'll comment on motivation for this limit.

rautenrieth-da reviewed (approved):
Thank you for deduplicating the deduplication logic 🙂
